### ISSUE TITLE
Deserialize variables in Values.fs

### ISSUE THREAD

**Issue by xperiandri** [fsprojects/FSharp.Data.GraphQL#327]
### Description

I propose to move variables deserialization into https://github.com/fsprojects/FSharp.Data.GraphQL/blob/b072222ff0bbe62869d951cf383dd200acd7ae47/src/FSharp.Data.GraphQL.Server/Values.fs#L105 and pass `Map<string, obj>`where `obj` can be either `System.Text.Json.JsonElement` or another `Map<string, obj>`.
Because in that function we have all the required info to correctly get value from JSON.

This way a scalar coerce function will always receive `System.Text.Json.JsonElement` and for objects options with converters can be applied.

@jberzy what do you think?

### Alternatives

Define an interface that allows to convert arbitrary JSON toke (i.e. JsonElemen to JToken) to a requested value type and pass it.

**Comment by xperiandri** [comment]
Or actually, deserialization can happen in `Executor.AsyncExecute`, after the plan is created and before it is executed.
At that point, we already know variables types. So we can pass a deserialization function to `Executor.AsyncExecute`

**Comment by njlr** [comment]
I recently hit this issue in a nasty way, so would appreciate more type-safety in this area. 

Some thoughts... 

I'm not sure if `System.Text.Json.JsonElement` is the best choice since it ties the user to a particular JSON implementation. Perhaps the library should provide its own JSON-like type?

Anyway, I was thinking what an implementation might look like. 

Taking the idea from `Thoth.Json`, we can have decoders like so:

```fsharp
type Decoder<'t> = JsonElement -> Result<'t, string>
```

The library could provide common decoders like `int` and `string`. 

To apply this concept, the user would first define an F# type for their input:

```fsharp
type FooInput =
  {
    Bar : int
    Qux : string list
    Baz : string option
  }
```

Then a corresponding decoder:

```fsharp
let fooInputDecoder : Decoder<FooInput> =
  (fun el ->
    result {
      let! bar = el |> Decode.property "bar" Decode.int
      let! qux = el |> Decode.property "qux" (Decode.list Decode.string)
      let! baz = el |> Decode.property "baz" (Decode.optional Decode.string)

      return
        {
          Bar = bar
          Qux = qux
          Baz = baz
        }
    })
```

_(Note that using reflection we can provide a default decoder for arbitrary types)_

Then the decoder would be passed to the GraphQL type definition:

```fsharp
let fooInputObjectDef =
  Define.InputObject<FooInput>
    (
      "FooInput",
      [
        Define.Input ("bar", Types.SchemaDefinitions.Int)
        Define.Input ("qux", Types.SchemaDefinitions.ListOf Types.SchemaDefinitions.String)
        Define.Input ("baz", Types.SchemaDefinitions.Nullable Types.SchemaDefinitions.String)
      ],
      fooInputDecoder
    )
```

I think it would be quite easy for `Values.fs` to find the decoder - or it could run elsewhere. 

Any thoughts?

**Comment by njlr** [comment]
I had a go here: https://github.com/fsprojects/FSharp.Data.GraphQL/pull/350

**Comment by njlr** [comment]
Implementation of "decoders" idea here https://github.com/fsprojects/FSharp.Data.GraphQL/pull/351

**Comment by xperiandri** [comment]
Implemented in #418
